ports/posix/iot_logging.c

```diff
@@ -0,0 +1,405 @@
+    #error "IOT_LOG_WARN must be 2."
+#endif
+#if IOT_LOG_INFO != 3
+    #error "IOT_LOG_INFO must be 3."
+#endif
+#if IOT_LOG_DEBUG != 4
+    #error "IOT_LOG_DEBUG must be 4."
+#endif
+
+/**
+ * @def IotLogging_Puts( message )
+ * @brief Function the logging library uses to print a line.
+ *
+ * This function can be set by using a define. By default, the standard library
+ * [puts](http://pubs.opengroup.org/onlinepubs/9699919799/functions/puts.html)
+ * function is used.
+ */
+#ifndef IotLogging_Puts
+    #define IotLogging_Puts    puts
+#endif
+
+/*
+ * Provide default values for undefined memory allocation functions based on
+ * the usage of dynamic memory allocation.
+ */
+#if IOT_STATIC_MEMORY_ONLY == 1
+    /* Static memory allocation header. */
+    #include "iot_static_memory.h"
+
+/**
+ * @brief Allocate a new logging buffer. This function must have the same
+ * signature as [malloc](http://pubs.opengroup.org/onlinepubs/9699919799/functions/malloc.html).
+ */
+    #ifndef IotLogging_Malloc
+        #define IotLogging_Malloc    Iot_MallocMessageBuffer
+    #endif
+
+/**
+ * @brief Free a logging buffer. This function must have the same signature
+ * as [free](http://pubs.opengroup.org/onlinepubs/9699919799/functions/free.html).
+ */
+    #ifndef IotLogging_Free
+        #define IotLogging_Free    Iot_FreeMessageBuffer
+    #endif
+
+/**
+ * @brief Get the size of a logging buffer. Statically-allocated buffers
+ * should all have the same size.
+ */
+    #ifndef IotLogging_StaticBufferSize
+        #define IotLogging_StaticBufferSize    Iot_MessageBufferSize
+    #endif
+#else /* if IOT_STATIC_MEMORY_ONLY == 1 */
+    #ifndef IotLogging_Malloc
+        #include <stdlib.h>
+        #define IotLogging_Malloc    malloc
+    #endif
+
+    #ifndef IotLogging_Free
+        #include <stdlib.h>
+        #define IotLogging_Free    free
+    #endif
+#endif /* if IOT_STATIC_MEMORY_ONLY == 1 */
+
+/**
+ * @brief A guess of the maximum length of a timestring.
+ *
+ * There's no way for this logging library to know the length of a timestring
+ * before it's generated. Therefore, the logging library will assume a maximum
+ * length of any timestring it may get. This value should be generous enough
+ * to accommodate the vast majority of timestrings.
+ *
+ * @see @ref platform_clock_function_gettimestring
+ */
+#define MAX_TIMESTRING_LENGTH    ( 64 )
+
+/**
+ * @brief The longest string in #_pLogLevelStrings (below), plus 3 to accommodate
+ * `[]` and a null-terminator.
+ */
+#define MAX_LOG_LEVEL_LENGTH     ( 8 )
+
+/**
+ * @brief How many bytes @ref logging_function_genericprintbuffer should output on
+ * each line.
+ */
+#define BYTES_PER_LINE           ( 16 )
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Lookup table for log levels.
+ *
+ * Converts one of the @ref logging_constants_levels to a string.
+ */
+static const char * const _pLogLevelStrings[ 5 ] =
+{
+    "",      /* IOT_LOG_NONE */
```

Comment:
Empty string is not needed as IOT_LOG_NONE is not printed.

Reply:
Removed